waiter/src/waiter/interstitial.clj

```diff
@@ -0,0 +1,238 @@
+                       (assoc-in [:service-id->interstitial-promise service-id] new-interstitial-promise)))
+                 interstitial-state))
+             (swap! interstitial-state-atom))
+        (if-let [interstitial-promise (service-id->interstitial-promise interstitial-state-atom service-id)]
+          (do
+            (when (identical? new-interstitial-promise interstitial-promise)
+              (counters/inc! (metrics/waiter-counter "interstitial" "promise" "total"))
+              (log/error "processing interstitial promise for" service-id)
+              (try
+                (process-interstitial-promise interstitial-promise)
+                (catch Exception e
+                  (log/error e "error in processing interstitial promise")
+                  (resolve-promise! interstitial-promise :processing-error))))
+            interstitial-promise)
+          (do
+            (log/info "looping as interstitial promise was removed" {:iteration iteration :service-id service-id})
+            (recur (inc iteration) (promise)))))))
+
+(defn remove-resolved-interstitial-promises!
+  "Removes all resolved promises mapped against services listed in service-ids.
+   It returns the services from service-ids which no longer have promises mapped against them in the interstitial state."
+  [interstitial-state-atom service-ids]
+  (->> (fn [{:keys [service-id->interstitial-promise] :as interstitial-state}]
+         (->> service-ids
+              (filter #(realized? (service-id->interstitial-promise %)))
+              (apply dissoc service-id->interstitial-promise)
+              (assoc interstitial-state :service-id->interstitial-promise)))
+       (swap! interstitial-state-atom))
+  (->> @interstitial-state-atom
+       :service-id->interstitial-promise
+       keys
+       set
+       set/difference service-ids))
+
+(defn process-interstitial-secs
+  "Resolves the interstitial-promise after a delay on interstitial-secs seconds."
+  [service-id interstitial-secs interstitial-promise]
+  (if (pos? interstitial-secs)
+    (async/go
+      (-> interstitial-secs t/seconds t/in-millis async/timeout async/<!)
+      (log/info "interstitial timed out for" service-id)
+      (resolve-promise! interstitial-promise :interstitial-time-out))
+    (do
+      (log/info service-id "opted out of interstitial")
+      (resolve-promise! interstitial-promise :interstitial-opt-out))))
+
+(defn- process-scheduler-messages
+  "Processes messages from the scheduler.
+   In particular, it resolves all promises for services which have healthy instances.
+   It also removes from the interstitial state any resolved promises for services which are no longer available."
+  [interstitial-state-atom service-id->service-description current-available-service-ids scheduler-messages]
+  (loop [[[message-type message-data] & remaining-scheduler-messages] scheduler-messages
+         scheduler-available-service-ids #{}]
+    (if message-type
```

Comment:
Would it be clearer to extract the data you need from messages, and process based upon that data?